### PR TITLE
[fix_bug] Entity delete with delete relations

### PR DESCRIPTION
**Pull request by nickAS21**
## Pull Request description

-  Entity delete with delete relations:
`if entity deletion error,  it is restored, but relations deleted`

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

**Comment by ashvayka** [comment]
Please add tests for this code. Then we will be able to merge.

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 1]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+        doDelete("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isBadRequest())
+                .andExpect(statusReason(containsString("The dashboard referenced by the device profiles cannot be deleted!")));
+
+        Dashboard afterErrorDeleteDashboard = doGet("/api/dashboard/" + dashboardIdStr,  Dashboard.class);
+        Assert.assertNotNull("Dashboard is not found!", afterErrorDeleteDashboard);
+        Assert.assertEquals("Dashboard after delete error is not equals origin!", savedDashboard, afterErrorDeleteDashboard);
+
+        EntityRelation afterErrorDeleteDashboardRelation = doGet(url, EntityRelation.class);
+        Assert.assertNotNull("Relation is not found!", foundRelation);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:559]: It seems that instead `foundRelation` you want to assert the `afterErrorDeleteDashboardRelation`

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 2]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+        Assert.assertEquals("Relation after delete error Dashboard is not equals origin!", foundRelation, afterErrorDeleteDashboardRelation);
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:560]: it's hard to follow what is `foundRelation`. I look that the better way is to compare with the original variable `relation`
like 
`Assert.assertEquals("Relation after delete error Dashboard is not equals origin!", relation, afterErrorDeleteDashboardRelation);`

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java (hunk 1 above)
Comment (smatvienko-tb) [reply application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:559]: Message `Relation is not found` on assertion is expected to be unique. It might be confusing to detect the problem by non-unique message. Please, add some sensible suffix to the message like 'Relation not found after dashboard deletion bad request'

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 3]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+                savedDashboard.getId(), alarmOperationResult.getPropagatedEntitiesList().get(0));
+
+        DeviceProfile deviceProfile =  createDeviceProfile();
+        deviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile savedDeviceProfile = doPost("/api/deviceProfile", createDeviceProfile(), DeviceProfile.class);
+
+        savedDeviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile updatedDeviceProfile = doPost("/api/deviceProfile", savedDeviceProfile, DeviceProfile.class);
+
+        String typeRelation = EntityRelation.CONTAINS_TYPE;
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:538]: I suggest testing the corner cases
1. no relations (probably already tested earlier)
2. one relation (your test)
3. many relations (two or more relations). from Tenant to Dashboard, with different Types or any you want

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 4]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", dashboardIdStr))));
+        doGet(url)
+                .andExpect(status().isNotFound())
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", foundRelation.getTo().getId().toString()))));
+        AlarmOperationResult afterSuccessDeleteDashboardAlarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                0, afterSuccessDeleteDashboardAlarmOperationResult.getPropagatedEntitiesList().size());
+    }
+
+    protected DeviceProfile createDeviceProfile() {
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:585]: the parent AbstractServiceTest already have method createDeviceProfile(). Please, use the existed method.
```
createDeviceProfile(tenantId, "my dev profile");
```

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 5]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
         Assert.assertEquals(0, pageData.getData().size());
     }
 
+    @Test
+    public void testDeleteDashboardExistsRelation_Error_RestoreRelation_DeleteRelation_DeleteDashboard_Ok() throws Exception {
+        Dashboard dashboard = new Dashboard();
+        dashboard.setTitle("My dashboard");
+        Dashboard savedDashboard = doPost("/api/dashboard", dashboard, Dashboard.class);
+
+        Alarm savedAlarm = createAlarm(savedDashboard.getId());
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:523]: Copy-paste the method `createAlarm()` from another class is not the best practice. 
Probably it is enough to call just alarmService.createOrUpdateAlarm(savedAlarm)

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 6]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+        doGet("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isNotFound())
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", dashboardIdStr))));
+        doGet(url)
+                .andExpect(status().isNotFound())
+                .andExpect(statusReason(containsString(msgErrorNoFound("Dashboard", foundRelation.getTo().getId().toString()))));
+        AlarmOperationResult afterSuccessDeleteDashboardAlarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                0, afterSuccessDeleteDashboardAlarmOperationResult.getPropagatedEntitiesList().size());
+    }
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:583]: the best practice is cleanup everything you created during this test

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java [hunk 7]
```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
...
+        deviceProfile.setTransportType(DeviceTransportType.DEFAULT);
+        deviceProfileData.setTransportConfiguration(new DefaultDeviceProfileTransportConfiguration());
+        deviceProfile.setProfileData(deviceProfileData);
+        deviceProfile.setDefault(false);
+        deviceProfile.setDefaultRuleChainId(null);
+        return deviceProfile;
+    }
+
+
+    private Alarm createAlarm(EntityId entityId) throws Exception {
```
Comment (smatvienko-tb) [inline application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:603]: code duplicate

File: dao/src/main/java/org/thingsboard/server/dao/tenant/TenantProfileServiceImpl.java [hunk 8]
```diff
@@ -112,7 +111,8 @@ public void deleteTenantProfile(TenantId tenantId, TenantProfileId tenantProfile
         this.removeTenantProfile(tenantId, tenantProfileId, false);
     }
 
-    private void removeTenantProfile(TenantId tenantId, TenantProfileId tenantProfileId, boolean isDefault) {
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/tenant/TenantProfileServiceImpl.java:115]: A private method is never responsible for a transaction. You need to find the public method that calls this private method and make those transactional:
image
image

image

File: dao/src/main/java/org/thingsboard/server/dao/user/UserServiceImpl.java [hunk 9]
```diff
@@ -201,6 +202,7 @@ public UserCredentials replaceUserCredentials(TenantId tenantId, UserCredentials
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/user/UserServiceImpl.java:205]: test required

File: dao/src/main/java/org/thingsboard/server/dao/tenant/TenantServiceImpl.java [hunk 10]
```diff
@@ -171,6 +171,7 @@ public Tenant saveTenant(Tenant tenant) {
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/tenant/TenantServiceImpl.java:174]: test required

File: dao/src/main/java/org/thingsboard/server/dao/rule/BaseRuleChainService.java [hunk 11]
```diff
@@ -748,6 +748,7 @@ private void deleteRuleNodes(TenantId tenantId, List<RuleNode> ruleNodes) {
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/rule/BaseRuleChainService.java:751]: BTW. check who is calling this method. there is a single call that is already transactional. But as it is already public, you can put transactional over here. Please, add the test

File: dao/src/main/java/org/thingsboard/server/dao/entityview/EntityViewServiceImpl.java [hunk 12]
```diff
@@ -294,6 +295,7 @@ public List<EntityView> findEntityViewsByTenantIdAndEntityId(TenantId tenantId,
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/entityview/EntityViewServiceImpl.java:298]: Test required

File: dao/src/main/java/org/thingsboard/server/dao/edge/EdgeServiceImpl.java [hunk 13]
```diff
@@ -182,6 +183,7 @@ public Edge unassignEdgeFromCustomer(TenantId tenantId, EdgeId edgeId) {
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/edge/EdgeServiceImpl.java:186]: Test required

File: dao/src/main/java/org/thingsboard/server/dao/device/DeviceProfileServiceImpl.java [hunk 14]
```diff
@@ -159,7 +159,8 @@ public void deleteDeviceProfile(TenantId tenantId, DeviceProfileId deviceProfile
         this.removeDeviceProfile(tenantId, deviceProfile);
     }
 
-    private void removeDeviceProfile(TenantId tenantId, DeviceProfile deviceProfile) {
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/device/DeviceProfileServiceImpl.java:162]: Private have to stay private. Please, find the public methods to use with transactions (see comment on tenant for details)

File: dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java [hunk 15]
```diff
@@ -110,6 +111,7 @@ public Customer saveCustomer(Customer customer) {
     }
 
     @Override
+    @Transactional
     public void deleteCustomer(TenantId tenantId, CustomerId customerId) {
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java:115]: Test required

File: dao/src/main/java/org/thingsboard/server/dao/asset/BaseAssetService.java [hunk 16]
```diff
@@ -150,6 +149,7 @@ public Asset unassignAssetFromCustomer(TenantId tenantId, AssetId assetId) {
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/asset/BaseAssetService.java:152]: test required

File: dao/src/main/java/org/thingsboard/server/dao/alarm/BaseAlarmService.java [hunk 17]
```diff
@@ -142,6 +143,7 @@ public PageData<AlarmData> findAlarmDataByQueryForEntities(TenantId tenantId,
     }
 
     @Override
+    @Transactional
```
Comment (smatvienko-tb) [inline dao/src/main/java/org/thingsboard/server/dao/alarm/BaseAlarmService.java:146]: test required

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Hey! It's nice to see the data consistency improvement.
Many entities have changed, so the main effort is to test all this stuff.
Please, keep in mind that some entities has `caches` and have to be safely maintained using the `redis` and `redis-cluster` caches. 
You can find the black-box-tests with Redis cache over here https://github.com/thingsboard/thingsboard/blob/master/msa/black-box-tests/README.md
Black box tests may not cover your use case, but you can improve it if needed.
Please, follow my comments below. 
I am looking forward to seeing your changes in the next review.

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java (hunk 1 above)
Comment (nickAS21) [reply application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:559]: Ok

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java (hunk 2 above)
Comment (nickAS21) [reply application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:560]: ` Assert.assertEquals("Found relation is not equals origin!", relation, foundRelation);`
It's the same

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java (hunk 5 above)
Comment (nickAS21) [reply application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java:523]: OK

File: dao/src/main/java/org/thingsboard/server/dao/device/DeviceProfileServiceImpl.java (hunk 14 above)
Comment (nickAS21) [reply dao/src/main/java/org/thingsboard/server/dao/device/DeviceProfileServiceImpl.java:162]: ok

File: application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java (hunk 3 above)
Comment (nickAS21) [reply application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java]: Ok

File: dao/src/main/java/org/thingsboard/server/dao/alarm/BaseAlarmService.java (hunk 17 above)
Comment (nickAS21) [reply dao/src/main/java/org/thingsboard/server/dao/alarm/BaseAlarmService.java:146]: removed

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
@nickAS21 , @ashvayka 
Sorry, but it's hard to read 450+ changes for the bugfix.
Not all changed services are tested. 
Not all services are fixed.
The test methods are long and hard to understand. And some more...

I spent some time developing an example of what I expect to see in PR
https://github.com/thingsboard/thingsboard/pull/7140
Hopefully, this will help

Please, consider following this example.
Have a good day!
